add custom and flexible DRF pagination to Asset model for admin etc

hburgund opened the pull request:
fixes #352

hburgund committed:
add custom and flexible DRF pagination to Asset model for admin etc

hburgund committed:
remove top-level wrappers from tags/ and uigroups/ response objects to be consistent and expected

hburgund committed:
change POST assets/ param name to envelope_ids to be consistent with PATCH assets/

hburgund committed:
generate POST assets/ error response if more than 1 envelope_id is passed; only one can be passed since related session and language are taken from envelope. additional envelope_ids can be added to asset later via PATCH assets/

hburgund commented:
includes partial work on #355 as well

IllyaMoskvin commented on roundware/api2/views.py:
```diff
@@ -73,8 +124,11 @@ def create(self, request):
         """
         if "file" not in request.data:
             raise ParseError("Must supply file for asset content")
+        if not request.data["envelope_ids"].isdigit():
```
Can we change this check to accept `envelope_ids` as an array, but with only one item? We can also check to ensure that item is a digit, similar to how it's done here.

If I understand correctly, with these changes, it expects `envelope_ids` to be an integer, rather than an array of integers. The parameter's name is consistent with the `envelope_ids` expected by the `PATCH` call, which is good, but the parameter type is different, which is bad.

Plus, when I see a plural variable name, I typically expect that variable's value to be an array. This is probably more of a personal stylistic convention than any sort of standard, but in that context, expecting `envelope_ids` to be an integer seems surprising.

If we'd rather not impose a single-length-array restriction on `envelope_ids`, I'd rather keep it as `envelope_id` here. Alternatively, we can expect an array, but silently recover and pick out only the first item, without raising a `ParseError`.

hburgund replied:
I agree 100% with you and spent a bunch of time trying to figure out how to accomplish this, but was unsuccessful. The primary way `POST` and `PATCH` are different, as best I can tell, is that `PATCH assets/:id/` takes data in `application/json` format to be consistent with the rest of the API, but `POST assets/` has to send its data in `multipart/form-data` because there is a binary attached (at least this is my understanding). 

I was unable to figure out how to send data in an array using `multipart/form-data` but I can't imagine that this isn't possible. I'm guessing that my problem was actually getting Postman to do this; it is probably no problem from your code, yes?

I have no problem imposing a single-item-array restriction similar to what I've already done with a  useful error message.

Per our conversation, I will experiment more with changing the way the server processes this data once I can figure out how to get Postman to actually send arrays properly.

hburgund replied:
### Some notes from Slack conversation with @IllyaMoskvin 

I've switched over to using cURL and am still having to do very hacky things to get anything to work. I am not able to find a consistent good approach to sending `form-data` parameter values with a binary file simultaneously. The data always ends up coming through not as a list/array but a string or some other type like unicode or something. I could convert this server-side, but this seems way more hacky than just doing it the way we have it now, despite the inconsistency.

I don't want to waste any more time on this without knowing exactly how Angular would ideally send this request through. Is there a way to replicate what Angular would send with a file as well as some array data so I can test?

perhaps just code your stuff as you'd like to and then I can grab your code and do some testing to see how the data comes through to the server and then make changes accordingly?

**imoskvin** 
I'm not sure that I'm sending any actual arrays... It could all be comma-separated strings.

**halsey**
so your current code sends the file along with `form-data` that includes parameters in array format?

I am trying to modify the code to accept either, but since I have no way of properly sending data as arrays, I do not know if that part works.

I would like to have a method for properly sending the request with the file and array data so I can see how to make the server handle it, but can't even get cURL to do it properly.

### Decision
Keep as is for now: 
* `envelope_ids` both input and output
* only one `envelope_id` allowed as input
* error displayed if more than one `envelope_id` posted
* input is comma-separated string of ids for all ManyToMany params in `form-data` requests

IllyaMoskvin commented on roundware/api2/views.py:
```diff
@@ -73,8 +124,11 @@ def create(self, request):
+                             "You can add more envelope_ids in subsequent PATCH calls")
         try:
-            result = add_asset_to_envelope(request, envelope_id=request.data["envelope_id"])
+            result = add_asset_to_envelope(request, envelope_id=request.data["envelope_ids"])
```
Please see above comment re: `envelope_ids`.

IllyaMoskvin reviewed (changes requested):
Everything's looking good, except for the `envelope_ids` value type inconsistency. Looking forward to testing out pagination on the frontend. :+1:

hburgund committed:
fix POST assets/ alt_text_loc_ids

hburgund committed:
add sorting by parameter to GET assets/ - for use in admin

hburgund committed:
remove mixin in favor of keeping assets list() inside viewset class

filter_backends is enabled by use of self.filter_queryset(self.get_queryset()) instead of AssetFilterSet(request.query_params); with this change in place, no need to use mixin, so this is a nice simplification

hburgund committed:
allow for both ordering and filtering to work on GET assets/;

this requires explicitly including DjangoFilterBackend as a filter_backend along with OrderingFilter despite DjangoFilterBackend being specified in SETTINGS
